### ISSUE TITLE
[libc++] Automatic `iterator_category` detection misbehaves when `::reference` is an rvalue reference, refuses to accept a forward iterator

### ISSUE THREAD

**Issue by HolyBlackCat** [llvm/llvm-project#57233]
Since C++20, `std::iterator_traits` can auto-detect `::iterator_category`, if the iterator doesn't set it.

Forward iterators require `::reference` (aka the return type of `operator*`) to be a reference (any reference), but libc++ only accepts lvalue references here.

Example:
```cpp
#include <iterator>
#include <type_traits>

template <typename T>
struct A
{
    using value_type = std::remove_cvref_t<T>;
    using difference_type = int;
    T operator*() const;
    A &operator++();
    A operator++(int);
    bool operator==(const A &) const;
};

// Ok.
static_assert(std::is_same_v<std::iterator_traits<A<int &>>::iterator_category, std::forward_iterator_tag>);
// Should pass but fails, the category is `std::input_iterator_tag`.
static_assert(std::is_same_v<std::iterator_traits<A<int &&>>::iterator_category, std::forward_iterator_tag>);
```

I blame this on cppreference, which used to incorrectly say that only lvalue references are allowed there. It was fixed since then.

See [`[forward.iterators]/1.3`](https://eel.is/c++draft/iterators#forward.iterators-1.3). Also see [LWG issue 1211](https://cplusplus.github.io/LWG/issue1211) (from 2009), which was resolved by [N3066](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2010/n3066.html) (in 2010).

This was discussed on SO: https://stackoverflow.com/q/73353152

[libstdc++](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106676) and MSVC's standard library share the exact same bug.

This was tested on Clang 14.0.0 and on trunk (16.0.0).

**Comment by frederick-vs-ja** [comment]
I don't think this is a bug of library implementation. Since `is_lvalue_reference_v` is explicitly used in [[iterator.traits]/2](https://eel.is/c++draft/iterator.traits#2) for _`cpp17-forward-iterator`_. If you want to fix such inconsistency, I think you should submit an LWG issue to change `is_lvalue_reference_v` in [iterator.traits]/2 to `is_reference_v`.

**Comment by HolyBlackCat** [comment]
I've mailed the issue.

**Comment by frederick-vs-ja** [comment]
FYI [LWG3798](https://cplusplus.github.io/LWG/issue3798) has been submitted.
